CorsixTH/Lua/app.lua

```diff
@@ -374,26 +375,38 @@ function App:init()
   return true
 end
 
---! Tries to initialize the user and built in level directories, returns true on
+--! Tries to initialize the user level and campaign directories, returns true on
 --! success and false on failure.
-function App:initUserLevelDir()
+-- TODO: Integrate other directory initialisations into this function
+function App:initUserDirectories()
   local conf_path = self.command_line["config-file"] or "config.txt"
-  self.user_level_dir = self.config.levels or
-      conf_path:match("^(.-)[^" .. pathsep .. "]*$") .. "Levels"
 
-  if self.user_level_dir:sub(-1, -1) == pathsep then
-    self.user_level_dir = self.user_level_dir:sub(1, -2)
-  end
-  if lfs.attributes(self.user_level_dir, "mode") ~= "directory" then
-    if not lfs.mkdir(self.user_level_dir) then
-      print("Notice: Level directory does not exist and could not be created.")
-      return false
+  -- Attempt to set the user's directory choice
+  -- param dir (path) The defined path of the folder by the user
+  -- param label (string) What folder is being set, for the error dialogue
+  -- return The fully qualified path
+  local function setUserDir(dir, label)
+    if dir:sub(-1, -1) == pathsep then
+      dir = dir:sub(1, -2)
     end
+    if lfs.attributes(dir, "mode") ~= "directory" then
+      if not lfs.mkdir(dir) then
+        print("Notice:" .. label .. "directory does not exist and could not be created.")
+        return false -- Force game exit
+      end
+    end
+    if dir:sub(-1, -1) ~= pathsep then
+     dir = dir .. pathsep
+    end
+    return dir
   end
-  if self.user_level_dir:sub(-1, -1) ~= pathsep then
-    self.user_level_dir = self.user_level_dir .. pathsep
-  end
-  return true
+
+  self.user_level_dir = self.config.levels or
+      conf_path:match("^(.-)[^" .. pathsep .. "]*$") .. "Levels"
+  self.user_level_dir = setUserDir(self.user_level_dir, "Levels")
+  self.user_campaign_dir = self.config.campaigns or
+      conf_path:match("^(.-)[^" .. pathsep .. "]*$") .. "Campaigns"
+  self.user_campaign_dir = setUserDir(self.user_campaign_dir, "Campaigns")
```

Comment:
Will condense more in a future PR

Reply:
Insert an empty line below 406 so it's easier to understand it's the same thing twice but with a different kind of directly.

Re-use of `self.usr_xyz_dir` variable is confusing, change the first assignment to a local variable.

If you stick the common parts in a function and supply the non-common parts as parameters, future PR would be pretty much done?

Reply:
That was the plan for the future PR, yes -- mainly to go with https://github.com/CorsixTH/CorsixTH/pull/2273#issuecomment-1287763941

Reply:
ok, fair enough.